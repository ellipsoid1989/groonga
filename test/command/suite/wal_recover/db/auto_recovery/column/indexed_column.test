# _database_close/_database_reopen can't use with HTTP because requests via HTTP are always
# processed by workers. Workers never use primary WAL role.
#@require-interface stdio

#@require-feature message_pack

#$GROONGA_ALLOW_DATABASE_REOPEN=yes
#$GRN_WAL_ROLE=primary

table_create Terms TABLE_PAT_KEY ShortText \
  --normalizer NormalizerNFKC130 \
  --default_tokenizer TokenNgram

table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content

load --table Data
[
{"_key": "abc",    "content": "a and b and c"},
{"_key": "ABC",    "content": "A and B and C"},
{"_key": "abcdef", "content": "a and b and c and d and e and f"}
]

delete --table Data --key ABC

dump

_database_close

#@copy-path #{db_path}.0000102 #{db_path}.0000102.wal
#@add-important-log-levels notice
_database_reopen
#@remove-important-log-levels notice

dump

index_column_diff --table Terms --name data_content

select Data \
  --match_columns 'content * 2' \
  --query d \
  --output_columns _id,_key,content,_score
