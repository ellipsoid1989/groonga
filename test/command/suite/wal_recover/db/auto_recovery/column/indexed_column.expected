table_create Terms TABLE_PAT_KEY ShortText   --normalizer NormalizerNFKC130   --default_tokenizer TokenNgram
[[0,0.0,0.0],true]
table_create Data TABLE_PAT_KEY ShortText
[[0,0.0,0.0],true]
column_create Data content COLUMN_SCALAR ShortText
[[0,0.0,0.0],true]
column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content
[[0,0.0,0.0],true]
load --table Data
[
{"_key": "abc",    "content": "a and b and c"},
{"_key": "ABC",    "content": "A and B and C"},
{"_key": "abcdef", "content": "a and b and c and d and e and f"}
]
[[0,0.0,0.0],3]
delete --table Data --key ABC
[[0,0.0,0.0],true]
dump
table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenNgram --normalizer NormalizerNFKC130

load --table Data
[
["_key","content"],
["abc","a and b and c"],
["abcdef","a and b and c and d and e and f"]
]

column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content
_database_close
[[0,0.0,0.0],true]
_database_reopen
[[0,0.0,0.0],true]
#|e| [reader][read-entry][ja][recover][Data.content] must be map: <positive-integer>(2): path:<db/db.0000102>
#|n| [db][wal][recover] rebuild broken column: <Data.content>(258)
#|n| DDL:260:column_create Data #recovering#content
#|n| spec:260:update:Data.#recovering#content:65(column:var_size):14(ShortText)
#|n| [db][wal][recover] rebuild broken column: <Terms.data_content>(259)
#|n| DDL:261:column_create Terms #recovering#data_content
#|n| spec:261:update:Terms.#recovering#data_content:72(column:index):257(Data)
#|n| DDL:261:set_source Terms.#recovering#data_content Data.#recovering#content
#|n| spec:260:update:Data.#recovering#content:65(column:var_size):14(ShortText)
#|n| DDL:2147483649:set_normalizers NormalizerNFKC130
#|n| DDL:2147483649:set_normalizers NormalizerNFKC130
#|n| spec:261:update:Terms.#recovering#data_content:72(column:index):257(Data)
#|n| DDL:258:obj_remove Data.content
#|n| DDL:259:obj_remove Terms.data_content
#|n| spec:259:remove:Terms.data_content:72(column:index)
#|n| spec:258:remove:Data.content:65(column:var_size)
#|n| [db][wal][recover] succeeded to rebuild broken column: <Data.content>(260)
#|n| [db][wal][recover] succeeded to rebuild broken column: <Terms.data_content>(261)
dump
table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenNgram --normalizer NormalizerNFKC130

load --table Data
[
["_key","content"],
["abc","a and b and c"],
["abcdef","a and b and c and d and e and f"]
]

column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content
index_column_diff --table Terms --name data_content
[[0,0.0,0.0],[]]
select Data   --match_columns 'content * 2'   --query d   --output_columns _id,_key,content,_score
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        1
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "content",
          "ShortText"
        ],
        [
          "_score",
          "Int32"
        ]
      ],
      [
        3,
        "abcdef",
        "a and b and c and d and e and f",
        2
      ]
    ]
  ]
]
